### PR TITLE
Deprecate setting shared cache on non-frozen nodes

### PR DESCRIPTION
**Pull request by jasontedor**
This commit deprecates the ability to set the shared cache to be positive on non-frozen nodes.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java [hunk 1]
```diff
@@ -67,9 +79,48 @@
...
+            public void validate(final ByteSizeValue value) {
+
+            }
+
+            @Override
+            public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+                if (value.getBytes() > 0) {
+                    @SuppressWarnings("unchecked") final List<DiscoveryNodeRole> roles =
+                        (List<DiscoveryNodeRole>) settings.get(NodeRoleSettings.NODE_ROLES_SETTING);
+                    final DiscoveryNode fake = new DiscoveryNode(
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:98]: Rather than constructing this fake `DiscoveryNode`, what about making `DataTier.isFrozenNode(...)` invoke a separate static method that takes a `Set<DiscoveryNodeRole>` so that method could be used directly? It feels weird to construct a whole fake node just for this purpose.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java [hunk 2]
```diff
@@ -67,9 +79,48 @@
...
+
+            @Override
+            public void validate(final ByteSizeValue value) {
+
+            }
+
+            @Override
+            public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+                if (value.getBytes() > 0) {
+                    @SuppressWarnings("unchecked") final List<DiscoveryNodeRole> roles =
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:96]: I'd feel better if we had an assert that `roles` was not null so we fail fast, or if we want to be lenient (which we probably don't) do `settings.get(NodeRoleSettings.NODE_ROLES_SETTING, Collections.emptyList())`

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java [hunk 3]
```diff
@@ -67,9 +79,48 @@
...
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = new Setting<>(
         SHARED_CACHE_SETTINGS_PREFIX + "size",
-        ByteSizeValue.ZERO,
+        ByteSizeValue.ZERO.getStringRep(),
+        s -> ByteSizeValue.parseBytesSizeValue(s, SHARED_CACHE_SETTINGS_PREFIX + "size"),
+        new Setting.Validator<ByteSizeValue>() {
+
+            @Override
+            public void validate(final ByteSizeValue value) {
+
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:90]: Should we disallow a `ByteSizeValue` of `-1b` since that is a valid `ByteSizeValue` object here?

It feels like the setting should be either 0b or > 0b, but not -1.

File: x-pack/plugin/searchable-snapshots/src/main/plugin-metadata/plugin-security.codebases [hunk 4]
```diff
@@ -0,0 +1 @@
+preallocate: org.elasticsearch.xpack.searchablesnapshots.preallocate.Preallocate
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/main/plugin-metadata/plugin-security.codebases:1]: This seems unrelated to this change?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java [hunk 5]
```diff
@@ -67,9 +79,48 @@
...
+                        Set.of(roles.toArray(DiscoveryNodeRole[]::new)), Version.CURRENT
+                    );
+                    if (DataTier.isFrozenNode(fake) == false) {
+                        deprecationLogger.deprecate(
+                            DeprecationCategory.SETTINGS,
+                            "shared_cache",
+                            "setting [{}] to be positive [{}] on node without the data_frozen role is deprecated, roles are [{}]",
+                            SHARED_CACHE_SETTINGS_PREFIX + "size",
+                            value.getStringRep(),
+                            roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.toUnmodifiableList())
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:111]: Perhaps
```suggestion
                            roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.joining(","))
```
to avoid the double `[[]]` wrapping

**Review by dakrone — APPROVED** [review]
Generally LGTM, but I did leave some comments and questions

File: x-pack/plugin/searchable-snapshots/src/main/plugin-metadata/plugin-security.codebases (hunk 4 above)
Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/plugin-metadata/plugin-security.codebases:1]: It's needed to be able to run the relevant unit tests inside of IntelliJ.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java (hunk 1 above)
Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java]: It felt weird to add a whole method just for this purpose. 🤷‍♀️

Especially since it's hard to come up with a good name for the method. `isFrozenNode` isn't great because it's not testing a node, `hasFrozenRole` isn't great because it can also include the `data` role and should return true.

I pushed 892f575e1234dc905cf67a1316ad24c738abca4e.

Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java]: It felt weird to add a whole method just for this purpose. 🤷‍♀️

Especially since it's hard to come up with a good name for the method. `isFrozenNode` isn't great because it's not testing a node, `hasFrozenRole` isn't great because it can also include the `data` role and should return true.

I pushed 892f575e1234dc905cf67a1316ad24c738abca4e.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java (hunk 2 above)
Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:100]: I don't think `roles` can be `null` here. We don't allow explicitly null-valued settings, since that represents unset, and therefore the default value.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java (hunk 3 above)
Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java:91]: I pushed 3f023831460b08397e1e56f5d2d82ba131d832b6.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheServiceTests.java [hunk 6]
```diff
@@ -192,6 +197,19 @@ public void testDecay() throws IOException {
...
+        DiscoveryNode.setAdditionalRoles(Set.of(
+            DataTier.DATA_HOT_NODE_ROLE, DataTier.DATA_WARM_NODE_ROLE, DataTier.DATA_COLD_NODE_ROLE, DataTier.DATA_FROZEN_NODE_ROLE));
+        final Settings settings = Settings.builder()
+            .put(FrozenCacheService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(), "500b")
+            .put(FrozenCacheService.SNAPSHOT_CACHE_REGION_SIZE_SETTING.getKey(), "100b")
+            .putList(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), DataTier.DATA_HOT_NODE_ROLE.roleName())
+            .build();
+        FrozenCacheService.SNAPSHOT_CACHE_SIZE_SETTING.get(settings);
+        assertWarnings("setting [" + FrozenCacheService.SNAPSHOT_CACHE_SIZE_SETTING.getKey()
+            + "] to be positive [500b] on node without the data_frozen role is deprecated, roles are [[data_hot]]");
```
Comment (dakrone) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheServiceTests.java:210]:
```suggestion
            + "] to be positive [500b] on node without the data_frozen role is deprecated, roles are [data_hot]");
```

**Review by dakrone — APPROVED** [review]
Still LGTM
